impl/client/src/main/javascript/web/pentaho/visual/base/view.js

```diff
@@ -729,6 +735,95 @@ define([
 
       // endregion
 
+      // region IActionTarget implementation
+      act: function(action) {
+        if(!action) throw error.argRequired("action");
+
+        action.execute(this, this._getActionController(action));
+
+        return action;
+      },
+
+      actAsync: function(action) {
+        if(!action) throw error.argRequired("action");
+
+        return action.executeAsync(this, this._getActionController(action));
+      },
+
+      __genericActionController: null,
+
+      _getActionController: function(action) {
+        return this.__genericActionController ||
+            (this.__genericActionController = this.__createGenericActionController());
+      },
+
+      __createGenericActionController: function() {
+        return {
+          init: this._onActionPhaseInit.bind(this),
+          will: this._onActionPhaseWill.bind(this),
+          "do": this._onActionPhaseDo.bind(this),
+          "finally": this._onActionPhaseFinally.bind(this)
+        };
+      },
+      // endregion
+
+      // region Actions
+      __emitActionPhase: function(action, phase, isFinal) {
+        var eventType = action.type.id;
+
+        // TODO: emitGenericAsync when action is async.
+
+        this._emitGeneric(action, eventType, phase, isFinal ? null : _emitActionKeyArgs);
```

Comment:
In principle, the action replaces the old event, for all uses. For that one, not sure. If, clearly, a DOM event triggered the action, then an action property would make sense to hold it. This would support async actions as well (once the DOM event is always accessible at window.event anyway).